src/Drupal/Commands/core/WatchdogCommands.php

```diff
@@ -51,25 +50,63 @@ class WatchdogCommands extends DrushCommands
      * @filter-default-field message
      * @return \Consolidation\OutputFormatters\StructuredData\RowsOfFields
      */
-    public function show($substring = '', $options = ['format' => 'table', 'count' => 10, 'severity' => self::REQ, 'type' => self::REQ, 'extended' => false])
+    public function show($substring = '', $options = ['format' => 'table', 'count' => 10, 'severity' => self::REQ, 'type' => self::REQ, 'extended' => false, 'tail' => false])
     {
         $where = $this->where($options['type'], $options['severity'], $substring);
         $query = Database::getConnection()->select('watchdog', 'w')
             ->range(0, $options['count'])
             ->fields('w')
             ->orderBy('wid', 'DESC');
-        if (!empty($where['where'])) {
-            $query->where($where['where'], $where['args']);
-        }
-        $rsc = $query->execute();
-        while ($result = $rsc->fetchObject()) {
-            $row = $this->formatResult($result, $options['extended']);
-            $table[$row->wid] = (array)$row;
+        if (!$tail) {
+            if (!empty($where['where'])) {
+                $query->where($where['where'], $where['args']);
+            }
+            $rsc = $query->execute();
+            while ($result = $rsc->fetchObject()) {
+                $row = $this->formatResult($result, $options['extended']);
+                $table[$row->wid] = (array)$row;
+            }
+            if (empty($table)) {
+                $this->logger()->notice(dt('No log messages available.'));
+                return;
+            } else if ($tail === false) {
+                return new RowsOfFields($table);
+            }
         }
-        if (empty($table)) {
-            $this->logger()->notice(dt('No log messages available.'));
+
+        if (empty($where['where'])) {
+            $where = [
+              'where' => 'wid > :wid',
+              'args' => [],
+            ];
         } else {
-            return new RowsOfFields($table);
+            $where['where'] .= " AND wid > ?";
+        }
+
+        $last_seen_wid = 0;
+        while (true) {
+            $where['args'][':wid'] = $last_seen_wid;
+            $query = Database::getConnection()->select('watchdog', 'w')
+                ->range(0, $options['count'])
+                ->fields('w')
+                ->orderBy('wid', 'DESC');
+            $query->where($where['where'], $where['args']);
+
+            $rsc = $query->execute();
+            while ($result = $rsc->fetchObject()) {
+                if ($result->wid > $last_seen_wid) {
+                  $last_seen_wid = $result->wid;
+                }
+                // TODO: how should the output string be built?
+                // Just printing stuff to show it doing something.
+                $row = $this->formatResult($result, $options['extended']);
+                print "\n  " . $row->wid . "\t";
```

Comment:
ack. @greg-1-anderson [gave me some pointers](https://app.slack.com/client/T06GX3JTS/C62H9CWQM/thread/C62H9CWQM-1622834281.067000) re. output in Slack. one idea was to create another command for the tail case. thoughts?

Reply:
It's super weird when a command returns RowsOfFields sometimes, and nothing (or some completely different type) at other times, but since `--tail` never returns at all, I guess it's not so bad that the behavior is inconsistent.

I'm trying to deprecated `$this->io()` and `$this->writeln()` et. al. in Robo; the better way is to add an OutputInterface as the first parameter. Then you'll get an object you can `$writeln` on.

Reply:
I've update the PR to use an injected OutputInterface object.